compiler/src/dotty/tools/dotc/transform/SuperPhase.scala

```diff
@@ -0,0 +1,628 @@
+    def transformTry(tree: Try)(implicit ctx: Context): Tree = tree
+    def transformSeqLiteral(tree: SeqLiteral)(implicit ctx: Context): Tree = tree
+    def transformInlined(tree: Inlined)(implicit ctx: Context): Tree = tree
+    def transformTypeTree(tree: TypeTree)(implicit ctx: Context): Tree = tree
+    def transformBind(tree: Bind)(implicit ctx: Context): Tree = tree
+    def transformAlternative(tree: Alternative)(implicit ctx: Context): Tree = tree
+    def transformUnApply(tree: UnApply)(implicit ctx: Context): Tree = tree
+    def transformValDef(tree: ValDef)(implicit ctx: Context): Tree = tree
+    def transformDefDef(tree: DefDef)(implicit ctx: Context): Tree = tree
+    def transformTypeDef(tree: TypeDef)(implicit ctx: Context): Tree = tree
+    def transformTemplate(tree: Template)(implicit ctx: Context): Tree = tree
+    def transformPackageDef(tree: PackageDef)(implicit ctx: Context): Tree = tree
+    def transformStats(trees: List[Tree])(implicit ctx: Context): List[Tree] = trees
+    def transformUnit(tree: Tree)(implicit ctx: Context): Tree = tree
+    def transformOther(tree: Tree)(implicit ctx: Context): Tree = tree
+
+    /** Transform tree using all transforms of current group (including this one) */
+    def transformAllDeep(tree: Tree)(implicit ctx: Context): Tree =
+      superPhase.transformTree(tree, 0)
+
+    /** Transform tree using all transforms following the current one in this group */
+    def transformFollowingDeep(tree: Tree)(implicit ctx: Context): Tree =
+      superPhase.transformTree(tree, idxInGroup + 1)
+
+    /** Transform single node using all transforms following the current one in this group */
+    def transformFollowing(tree: Tree)(implicit ctx: Context): Tree =
+      superPhase.transformNode(tree, idxInGroup + 1)
+
+    protected def singletonGroup = new SuperPhase(Array(this))
+
+    override def run(implicit ctx: Context): Unit =
+      singletonGroup.run
+  }
+
+  private type Transformer[-T, R] = (T, Context) => R
+  private type IndexedTransformer[-T, R] = MiniPhase => Transformer[T, R]
+
+  private val idNodeTransformer: Transformer[AnyRef, AnyRef] = (t, ctx) => t
+  private val idContextTransformer: Transformer[AnyRef, Context] = (t, ctx) => ctx
+}
+import SuperPhase._
+
+class SuperPhase(val miniPhases: Array[MiniPhase]) extends Phase {
+  import ast.tpd._
+
+  override val phaseName =
+    if (miniPhases.length == 1) miniPhases(0).phaseName
+    else miniPhases.map(_.phaseName).mkString("SuperPhase{", ", ", "}")
+
+  private val cpy: TypedTreeCopier = cpyBetweenPhases
+
+  /** Transform node using all phases in this group that have idxInGroup >= start */
+  def transformNode(tree: Tree, start: Int)(implicit ctx: Context) =
+    nodeTransformer(tree.tag)(start)(tree, ctx)
+
+  /** Transform full tree using all phases in this group that have idxInGroup >= start */
+  def transformTree(tree: Tree, start: Int)(implicit ctx: Context): Tree = {
+    val tag = tree.tag
```

Comment:
We could avoid the megamorphic dispatch here if we had something like:
```scala
abstract class Tree[-T >: Untyped](tag: TreeTag) ...
// ...
class Ident(...) extends Tree(Tag.Ident)
```

Reply:
Yes, but at the cost of one field per tree. It might still be worth it, though. One needs to experiment.